### PR TITLE
Add delete revisions functions, add new translations, add german and …

### PR DESCRIPTION
**Pull request by evwerkz**
Hi, your Plugin is very good and helpful. 
I missing some delete functions for single and all revisions by model. So, i added two functions: delete by single revision and delete all by model.

And i add new translations values and germany and russina translations.

### CODE REVIEW

File: formwidgets/RevisionHistory.php [hunk 1]
```diff
@@ -195,4 +195,29 @@ private function getDiff($fieldName, $oldValue, $newValue)
...
+
+    /**
+     * Find revision, check if exist and delete this
+     */
+    public function onDeleteRevisionById()
+    {
+        $revision = Revision::where('id', input('revision_id'))->first();
+        if ($revision) {
+            $revision->delete();
+            Flash::success('This changes successfully deleted!');
```
Comment (rubenvanerk) [inline formwidgets/RevisionHistory.php:207]:
```suggestion
            Flash::success('This changes successfully deleted!');
            return;
```

Comment (rubenvanerk) [reply formwidgets/RevisionHistory.php:207]: Add a return to prevent the warning Flash from triggering

File: formwidgets/RevisionHistory.php [hunk 2]
```diff
@@ -195,4 +195,29 @@ private function getDiff($fieldName, $oldValue, $newValue)
...
+    }
+
+    /**
+     * Delete all revision by model id
+     */
+    public function onDeleteAllRevisionsByModel()
+    {
+        if($id = $this->model->id) {
+            Revision::where('revisionable_id', $id)->delete();
+            Flash::success('All changes successfully deleted!');
```
Comment (rubenvanerk) [inline formwidgets/RevisionHistory.php:219]:
```suggestion
            Flash::success('All changes successfully deleted!');
            return;
```

Comment (rubenvanerk) [reply formwidgets/RevisionHistory.php:219]: Add a return to prevent the warning Flash from triggering

File: formwidgets/RevisionHistory.php [hunk 3]
```diff
@@ -195,4 +195,29 @@ private function getDiff($fieldName, $oldValue, $newValue)
...
+            Flash::success('This changes successfully deleted!');
+        }
+        Flash::warning('This revision could not be found!');
+    }
+
+    /**
+     * Delete all revision by model id
+     */
+    public function onDeleteAllRevisionsByModel()
+    {
```
Comment (rubenvanerk) [inline formwidgets/RevisionHistory.php:216]: Add a check for `readOnly` like in `onRevertHistory()`

File: formwidgets/RevisionHistory.php [hunk 4]
```diff
@@ -195,4 +195,29 @@ private function getDiff($fieldName, $oldValue, $newValue)
 
         return Diff::htmlDiff(e($oldValue), e($newValue));
     }
+
+    /**
+     * Find revision, check if exist and delete this
+     */
+    public function onDeleteRevisionById()
+    {
```
Comment (rubenvanerk) [inline formwidgets/RevisionHistory.php:203]: Add a check for `readOnly` like in `onRevertHistory()`

File: formwidgets/RevisionHistory.php (hunk 1 above)
Comment (rubenvanerk) [reply formwidgets/RevisionHistory.php:207]: Also replace the string with a language string

File: formwidgets/RevisionHistory.php (hunk 2 above)
Comment (rubenvanerk) [reply formwidgets/RevisionHistory.php:219]: Also replace the string with a language string

File: formwidgets/revisionhistory/partials/_revisionhistory.htm [hunk 5]
```diff
@@ -1,3 +1,38 @@
+<?php if(!empty($history)) {?>
```
Comment (rubenvanerk) [inline formwidgets/revisionhistory/partials/_revisionhistory.htm:1]: Hide the delete all history button when revision history has `readOnly: true`.

File: formwidgets/revisionhistory/partials/_revisionhistory.htm [hunk 6]
```diff
@@ -88,6 +123,11 @@ <h4 class="modal-title">
                                             data-request-success="$(this).closest('.modal').modal('hide');">
                                         <?= e(trans('samuell.revisions::lang.revision.restore_all')) ?>
                                     </button>
+                                    <button type="submit" class="btn btn-danger" data-request="onDeleteRevisionById"
```
Comment (rubenvanerk) [inline formwidgets/revisionhistory/partials/_revisionhistory.htm:126]: The added button pushes the cancel button to a new line for the at least the dutch language. Increasing the width of the model could solve this.
image

**Review by rubenvanerk — COMMENTED** [review]
I think this is a great addition! I have added some comments to improve your PR

File: formwidgets/RevisionHistory.php [hunk 7]
```diff
@@ -214,10 +216,20 @@ public function onDeleteRevisionById()
...
+            return;
         }
         Flash::warning('This model could not be found!');
     }
+
+    private function ifReadOnlyChecker()
+    {
+        if ($this->readOnly) {
+            Flash::error(Lang::get('samuell.revisions::lang.revision.read_only_error'));
+            return;
```
Comment (rubenvanerk) [inline formwidgets/RevisionHistory.php:232]: Just returning would not be enough. The code after `$this->readOnlyChecker()` will still be executed.
Maybe throwing an exception could solve this problem?

**Comment by evwerkz** [comment]
Hi, i edited some functions and also the comments you made.
And i added an ajax popup function.

**Comment by Samuell1** [comment]
Hey amazing PR! Can you please split translations to different PR?

**Comment by rubenvanerk** [comment]
The ajax popup is a very good addition! My last couple of points:

- [ ] The diff styling is missing in the ajax popup.
- [ ] I'm not seeing the flash message when I'm trying to delete a revision and the widget is set to readOnly (removing the return in `returnFlashMessage` seems to fix this)
- [ ] I also don't really see the need for the `returnFlashMessage`. Following notation should work just fine:
```php
Flash::error(Lang::get('samuell.revisions::lang.revision.read_only_error'));
return;
```

File: formwidgets/revisionhistory/partials/_revisionhistory.htm (hunk 6 above)
Comment (evwerkz) [reply formwidgets/revisionhistory/partials/_revisionhistory.htm]: > The added button pushes the cancel button to a new line for the at least the dutch language. Increasing the width of the model could solve this.
> image

image

what about this? 
i add <div><br></div>

**Comment by evwerkz** [comment]
> The ajax popup is a very good addition! My last couple of points:
> 
> * [*]  The diff styling is missing in the ajax popup.
> * [*]  I'm not seeing the flash message when I'm trying to delete a revision and the widget is set to readOnly (removing the return in `returnFlashMessage` seems to fix this)
> * [*]  I also don't really see the need for the `returnFlashMessage`. Following notation should work just fine:
> 
> ```
> Flash::error(Lang::get('samuell.revisions::lang.revision.read_only_error'));
> return;
> ```

1. See top comment
2 & 3. Ok i delete returnFlashMessage Funktion and add you code:

```
Flash::error(Lang::get('samuell.revisions::lang.revision.read_only_error'));
return;
```

File: formwidgets/revisionhistory/partials/_revisionhistory.htm (hunk 6 above)
Comment (evwerkz) [reply formwidgets/revisionhistory/partials/_revisionhistory.htm]: image

German words are bigger.

width of the model: October habe some ajax popup style for width like this:

`data-size="large" - Popup size, available sizes: giant, huge, large, small, tiny`

https://octobercms.com/docs/ui/popup

File: formwidgets/RevisionHistory.php [hunk 8]
```diff
@@ -173,26 +168,70 @@ private function validateInput()
...
         return Diff::htmlDiff(e($oldValue), e($newValue));
     }
+
+    /**
+     * Find revision, check if exist and delete this
+     */
+    public function onDeleteRevisionById()
+    {
+        if ($this->readOnly) {
+            Flash::success(Lang::get('samuell.revisions::lang.revision.read_only_error'));
```
Comment (rubenvanerk) [inline formwidgets/RevisionHistory.php:198]: Please change this to an error flash message

**Review by rubenvanerk — CHANGES_REQUESTED** [review]
Sorry, but what do you mean with 'See top comment'?
What I meant with diff styling is that in the ajax popup, the styling is missing like in the history list:
image

File: formwidgets/RevisionHistory.php (hunk 8 above)
Comment (evwerkz) [reply formwidgets/RevisionHistory.php:198]: > Sorry, but what do you mean with 'See top comment'?
> What I meant with diff styling is that in the ajax popup, the styling is missing like in the history list:
> image

Oh my mistake. I thought you meant width styles for popup.

Comment (evwerkz) [reply formwidgets/RevisionHistory.php]: @rubenvanerk 
Is done. I added class "revision history" to popup partial.

**Comment by rubenvanerk** [comment]
Awesome, thanks for the effort! Can you fix the conflict so this PR can be merged?

@Samuell1 can you review and merge this PR?

**Comment by evwerkz** [comment]
> Awesome, thanks for the effort! Can you fix the conflict so this PR can be merged?
> 
> @Samuell1 can you review and merge this PR?

@rubenvanerk @Samuell1 

i hope, i did all right. On master are new functions: `getRelationDiff `and `getBooleanDiff`

**Comment by rubenvanerk** [comment]
LGTM!

**Comment by Samuell1** [comment]
Amazing PR! I dont have time now to review it fully but i trust you guys. 

I just need that version.yaml isnt edited yet and i can merge this. I release this later as 1.2.0 version with all changes.

**Pull request merged**